native/swift/Sources/wordpress-api/SafeRequestExecutor.swift

```diff
@@ -37,47 +37,58 @@ public final class WpRequestExecutor: SafeRequestExecutor {
 
     public func execute(_ request: WpNetworkRequest) async -> Result<WpNetworkResponse, RequestExecutionError> {
         do {
-            var urlrequest = request.asURLRequest()
-
-            // Set the user agent before `additionalHttpHeadersForAllRequests` so that it can be overridden that way
-            urlrequest.setValue(self.userAgent, forHTTPHeaderField: "User-Agent")
-
-            for (key, value) in additionalHttpHeadersForAllRequests {
-                urlrequest.addValue(value, forHTTPHeaderField: key)
-            }
-
-            let (data, response) = try await self.fetch(request: urlrequest)
-
+            let (data, response) = try await self.fetch(request: self.preflight(request.asURLRequest()))
             return .success(try WpNetworkResponse(data: data, request: request, response: response))
         } catch {
-            if errorIsHttpsError(error) {
-                return handleHttpsError(error, for: request)
-            }
-
-            if errorIsNonExistentSiteError(error) {
-                return handleNonExistentSiteError(error, for: request)
-            }
-
-            if errorIsDeviceIsOffline(error) {
-                return handleDeviceIsOfflineError(error)
-            }
+            return self.handleRequestError(error, request: request)
+        }
+    }
 
-            return .failure(.RequestExecutionFailed(
-                statusCode: nil,
-                redirects: nil,
-                reason: .genericError(errorMessage: error.localizedDescription)
-            ))
-       }
+    public func uploadMedia(mediaUploadRequest: MediaUploadRequest) async throws -> WpNetworkResponse {
+        let urlrequest = try await mediaUploadRequest.asUrlRequest()
+        let (_, response) = try await self.fetch(request: self.preflight(urlrequest))
```

Comment:
[`URLSession.uploadTask`](https://developer.apple.com/documentation/foundation/urlsession/uploadtask(with:from:completionhandler:)) is probably more appropriate for uploading media. We can choose to stream from a file or memory. Things may go wrong when users choose to upload a video if we always read the file content into memory.

Reply:
We're not currently reading the file into memory – instead, we make a file that contains the entire request body then provide that as an `InputStream` on the `URLRequest`. 

It might make sense to separate binary upload from the file metadata though, in which case we could definitely use an upload task